Framework/Indexing/inc/MantidIndexing/IndexInfo.h

```diff
@@ -0,0 +1,79 @@
+#ifndef MANTID_INDEXING_INDEXINFO_H_
+#define MANTID_INDEXING_INDEXINFO_H_
+
+#include "MantidIndexing/DllConfig.h"
+#include "MantidKernel/cow_ptr.h"
+
+#include <functional>
+#include <set>
+#include <vector>
+
+namespace Mantid {
+
+using specnum_t = int32_t;
+using detid_t = int32_t;
+
+namespace Indexing {
+
+/** IndexInfo is an object for holding information about spectrum numbers and
+  detector IDs associated to the spectra in a workspace.
+
+  @author Simon Heybrock
+  @date 2016
+
+  Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_INDEXING_DLL IndexInfo {
+public:
+  explicit IndexInfo(const size_t globalSize);
+  IndexInfo(std::vector<specnum_t> &&spectrumNumbers,
+            std::vector<std::vector<detid_t>> &&detectorIDs);
+  IndexInfo(
+      const size_t globalSize,
+      std::function<specnum_t(const size_t)> getSpectrumNumber,
+      std::function<const std::set<specnum_t> &(const size_t)> getDetectorIDs);
+
+  IndexInfo(const IndexInfo &other);
+
+  size_t size() const;
+
+  specnum_t spectrumNumber(const size_t index) const;
+  std::vector<detid_t> detectorIDs(const size_t index) const;
+
+  void setSpectrumNumbers(std::vector<specnum_t> &&spectrumNumbers) & ;
+  void setDetectorIDs(const std::vector<detid_t> &detectorIDs) & ;
+  void setDetectorIDs(std::vector<std::vector<detid_t>> &&detectorIDs) & ;
+
+private:
+  bool m_isLegacy{false};
```

Comment:
Could definitely use some documentation in `IndexInfo` about this `m_isLegacy` . I agree that it is currently needed, but this flag binds together quite a few other methods in the context of the MatrixWorkspace that could use some explanation for future readers.